verifier/src/report_body.rs

```diff
@@ -2,72 +2,95 @@
 
 //! Verifiers which operate on the [`ReportBody`]
 
-use crate::{And, AndError, VerificationError, Verifier};
+use crate::{AlwaysTrue, AndError, VerificationError, Verifier};
 use mc_sgx_core_types::{Attributes, ConfigId, ReportBody};
 use subtle::CtOption;
 
 /// Verify the report body is as expected.
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Default)]
+pub struct ReportBodyVerifier {
+    attributes_verifier: ReportBodyVerifierKind,
+    config_id_verifier: ReportBodyVerifierKind,
+}
+
 #[derive(Clone, Debug, Eq, Hash, PartialEq)]
-pub struct ReportBodyVerifier<'a> {
-    expected_report_body: ReportBody,
-    report_body: &'a ReportBody,
+pub enum ReportBodyVerifierKind {
+    AttributesVerifier(AttributesVerifier),
+    ConfigIdVerifier(ConfigIdVerifier),
+    AlwaysTrue(AlwaysTrue),
 }
 
-impl<'a> ReportBodyVerifier<'a> {
-    /// Create a new [`ReportBodyVerifier`] instance.
-    ///
-    /// # Arguments:
-    /// * `expected_report_body` - The expected report body.
-    /// * `report_body` - The report body to verify conforms to the
-    ///   `expected_report_body`.
-    pub fn new(expected_report_body: ReportBody, report_body: &'a ReportBody) -> Self {
-        Self {
-            expected_report_body,
-            report_body,
+impl Verifier<ReportBody> for ReportBodyVerifierKind {
+    type Error = VerificationError;
+
+    fn verify(&self, evidence: &ReportBody) -> CtOption<Self::Error> {
+        match self {
+            Self::AttributesVerifier(verifier) => verifier.verify(evidence),
+            Self::ConfigIdVerifier(verifier) => verifier.verify(evidence),
+            Self::AlwaysTrue(verifier) => verifier.verify(evidence),
```

Comment:
An off the cuff suggestion in https://github.com/mobilecoinfoundation/attestation/pull/3#discussion_r1110210689 was to have an accessor trait, but I'm thinking that mixes multiple abstraction levels.
My thought is the higher level logic that says to "Verify the app report body" will know where the app report body is and can pass it to the "ReportBodyVerifier" and it will know where the QE report body is and pass it to another "ReportBodyVerifier" instance. I'm thinking `QuoteVerifier`'s `verify()` will be provided a Quote and then it will pass the individual report bodies to the appropriate verifiers.

Having said what I said above, now that I'm looking at this I'm wondering if this should be:
```suggestion
            Self::AttributesVerifier(verifier) => verifier.verify(evidence.attributes()),
            Self::ConfigIdVerifier(verifier) => verifier.verify(evidence.config_id()),
            Self::AlwaysTrue(verifier) => verifier.verify(evidence),
```

Reply:
I've made the change in #12. 
I started adding in another member of ReportBody and it started to feel clunky having to initialize a ReportBody to use the verifier.

Reply:
Moved this pr to use an Accessor trait.